Add OpenFX OCIO plugin framework and examples

michdolan opened the pull request:
This PR includes a working OCIOColorSpace OFX plugin and a lower-level framework to support other OCIO plugins fairly simply. Each plugin creates a transform and passes it to an OCIOProcessor base class which manages the rendering on multiple threads using an OCIO::PackedImageDesc. This currently only supports the CPU renderer but could be extended to support the GPU renderer. All bit-depths are supported.

A large component of this work is on the CMake side, since it uses the OpenFX Support Library (C++ wrapper) for interacting with the C API. I initially tried to build a static lib from that source and link it into plugins, but the plugins were not loaded by hosts. The issue could have certainly been on my end, so open to ideas there. The current approach treats the support library as additional source and compiles it directly with the plugin. The Resolve OFX example _seems_ to use the same approach (by appearances at least), but i'll need to build it to confirm.

We can add more plugins in later PRs, and the overall UX of these will no doubt be tweaked as the UX working group iterates on implementation guidelines.

Resolve:
![ocio_ofx_controls](https://user-images.githubusercontent.com/2299963/114790328-96ee2280-9d52-11eb-839c-729dcab37f0c.png)
![ocio_ofx_button](https://user-images.githubusercontent.com/2299963/114790760-4d520780-9d53-11eb-94ab-0ce01a72c5bd.png)
Natron:
![ocio_ofx_natron_controls](https://user-images.githubusercontent.com/2299963/114790353-a0778a80-9d52-11eb-960c-597d257e78a7.png)
![ocio_ofx_natron_button](https://user-images.githubusercontent.com/2299963/114790774-517e2500-9d53-11eb-9692-4470fbddaa58.png)

Signed-off-by: Michael Dolan <michdolan@gmail.com>

hodoulp commented on .github/workflows/ci_workflow.yml:
```diff
@@ -58,12 +58,13 @@ jobs:
-          # VFX CY2021
+          # GCC, VFX CY2021
           # -------------------------------------------------------------------
           - build: 12
```
[Question] Each entry in the matrix must now set several variables, and the number of variables is growing. In order to reduce the number of variables per entry to only the 'changed' ones, is there a way to have some global default values?

michdolan replied:
Good question. I'll have to do some investigation. We're using the GHA matrix in a somewhat unconventional way currently, mostly because the built-in matrix would result in too many permutations.

hodoulp commented on vendor/openfx/CMakeLists.txt:
```diff
@@ -0,0 +1,129 @@
+set(OFX_CXX_STANDARD ${CMAKE_CXX_STANDARD})
+if(${CMAKE_CXX_STANDARD} GREATER_EQUAL 17)
+    set(OFX_CXX_STANDARD 11)
+endif()
```
That could not be here. For example the OCIO library compiles in C++17 but openfx compiles in C++11. That could be problematic even if we were careful in the OCIO public API. This constraints must be moved in the global CMakeLists.txt as a check when openfx is ON.

remia replied:
Would it be worth maintaining a patch to have the OFX Support library C++17 compliant ?

michdolan replied:
I was thinking something similar. I have a hunch that most of the changes would be pretty boilerplate (i.e. replace `auto_ptr` with `unique_ptr`). I'll do some investigation into this. The Support Library source appears to have the same BSD 3-Clause license as OCIO, but ideally updates would be contributed back to openfx in the long term.

michdolan replied:
See note below, but I have removed the CMake commands in question here

remia commented:
Great work @michdolan, I tested that on Resolve macOS and it worked fine.

I'm not sure where you issues of having the support layer as a static lib comes from, here it loaded just fine with that setup. Maybe I missed something though, or it's a linker behaviour difference between platforms. On other note, I had to use a custom RPATH for the plugin pointing to the absolute installation library folder, for better or worse but it work when you copy the plugin to Resolve expected folder and install OCIO to a non standard location...

A bit curious about the GPU implementation, on Resolve at least, the OpenGL processing suite doesn't seems to be supported which probably means per frame upload / download of texture ? OpenCL / Metal / CUDA only seems officially supported. Also didn't really understood if that is a Resolve issue, but the multithreading seems quite disappointing.

remia commented on vendor/openfx/OCIOColorSpace.cpp:
```diff
@@ -0,0 +1,126 @@
+void OCIOColorSpace::render(const OFX::RenderArguments & args)
+{
+    // Get images
+    std::auto_ptr<OFX::Image> dst(dstClip_->fetchImage(args.time));
```
Use std::unique_ptr instead ?

remia commented on vendor/openfx/OCIOProcessor.h:
```diff
@@ -0,0 +1,39 @@
+    void setTransform(OCIO::ConstTransformRcPtr tr);
+
+    /* Process image on multiple threads */
+    void multiThreadProcessImages(OfxRectI procWindow);
```
Can add `override` qualifier.

michdolan commented:
@remia I think the OpenCL/CUDA/Metal support is specific to Resolve. I'm guessing they are maintaining their own version of: https://github.com/ofxa/openfx/blob/master/Support/Plugins/include/ofxsProcessing.H .

I agree on the multithreading performance. It's possible it could also be related to my choice of flags for the openfx interface.

michdolan commented:
I added the OpenFX Support Library source, modified to fix compiler errors and warnings for C++ 11, 14, and 17. I would like to contribute these changes back to openfx, but in the meantime I am proposing we house the modified code here, since it is BSD licensed, like OCIO. We _could_ also add a OpenGL or DirectX render extension like Resolve did with the other GPU APIs...

@jmertic for license check to make sure this is ok for now.

michdolan commented:
New OCIODisplay node:
![ocio_ofx_display](https://user-images.githubusercontent.com/2299963/115724091-a111b180-a34e-11eb-856f-fb40719e42b7.png)

Updated OCIOColorSpace and added warning for scene/display color space conversion:
![ocio_ofx_colorspace_warning](https://user-images.githubusercontent.com/2299963/115724374-e33af300-a34e-11eb-951b-5ddae8a53827.png)

This work intended to propose solution for #1369

michdolan commented:
![ocio_ofx_context](https://user-images.githubusercontent.com/2299963/116002549-39669b00-a5c8-11eb-856f-b2cfaab42ea4.png)

Added context override params like Nuke's "Context" tab. I took a slightly different approach and am creating a dedicated param per environment variable declared in the OCIO config "environment" section (vs. having a preset number of generic key/value pair params). Open to discussion as to whether this is the right approach. OFX handles this well in my tests of loading a different config in which the generated context params would differ due to different environment variables being declared (and thus dropping previously stored values). Tooltips provide a declared default (as pictured here).

zachlewis commented:
Hmmm... interesting! I really like this approach, and it rewards declaring the "environment" value; however, I'd still like to be able to set (and persist) values and expressions for Context variables that might not exist in the config; or for configs that don't define an environment section. 

For example, we have "Dynamic ColorSpace" and "Dynamic Display" nodes whose first two context (key, value) pairs are hard-coded with nuke tcl expressions to parse $SHOT and $SEQUENCE values from regex matching "input/filename" metadata. I'd like to assign / override those keys and values independently of the config. I don't know if Resolve has an equivalent concept (but I'd be really curious to find out). 

Of course, maybe there are better places for overriding those context variables defaults than in this node's interface...

Can OFX nodes inherit arbitrary metadata from each other? Imagine an OCIO Context node for tracking and overriding the environment on a scope local to downstream nodes; and all other nodes can use the proposed node-level, dynamic per-config overrides.

(And the idea of OCIO-node metadata would extend nicely for things like tracking image state / encoding, which could inform the default modes of Grading and ExposureContrast nodes... all sorts of stuff. just a thought)


Thanks Michael, I can't tell you how awesome these OFX nodes are.

michdolan commented:
![ocio_ofx_context2](https://user-images.githubusercontent.com/2299963/116025405-66de3380-a61e-11eb-9c75-d0a64fcb08c9.png)

Thanks for the excellent feedback @zachlewis ! Here's the other approach we could do, more like what Nuke is doing. Good point on the use case you described. I agree that seems like it may be the better way to go.

As for externalizing the context overrides to its own node (and passing via metadata), I agree that would be very DAG/node graph friendly in terms of workflow. I'll do some investigation to see if openfx provides that level of interaction with the host application. I know it could be done in Nuke by writing a custom Op type, but I'm not sure how aware openfx plugins are of the surrounding graph implementation.

michdolan commented:
![ocio_ofx_swap_cs](https://user-images.githubusercontent.com/2299963/116879103-397d2100-abee-11eb-9a86-1c9235fd8868.png)

Added button to swap color spaces.

Also, I did more testing on the dynamic context variable interface (as discussed in UX WG meeting) and the persistence of the parameter values through config changes with different declared environment variables. In Resolve the values are preserved, but not in Natron, so it seems implementation dependent. As a solution I can add a hidden param to preserve all context variables and populate them as the dynamic params are available.

michdolan commented:
Context overrides are now preserved through config changes with the dynamic context param UX. A secret param serializes all context overrides when any context param is modified. I have tested this in Resolve and Natron and both re functioning.

michdolan commented:
The last thing I'm going to do in this PR is support preservation of choice param values even when not present in the config. This will result in an error on loading a scene with a different OCIO config as opposed to resetting params silently.

zachlewis commented:
Thank you Michael -- preservation of choice param values even when not present in the config is exactly what I was hoping for. You're the best.

hodoulp commented:
@michdolan let us know when you think the work is completed so we can merge it. As this pull request is already 4-weeks old, I would suggest to merge it (if ready) to give a chance to the community to also test it.

doug-walker reviewed (changes requested):
This is so awesome, thank you Michael!

Like Remi, I had to copy the ofx bundle and mess around with the rpath to get it to work, but the plug-ins then worked for me in Flame.  There are a few issues as I mention in the code, the biggest being the bit-depth issue and the fact that the context var UI is not showing up.  But this is a huge feature for OCIO and very nicely implemented!

doug-walker commented:
I found another issue.  When I use a config that has a different set of views for each display, I'm noticing that when I change the display, the view menu doesn't update with the appropriate new values.  It still has the old values.  Does it update correctly in the other hosts you tested in?

michdolan commented:
Thanks for the feedback @doug-walker ! I'll work on refinements over the next couple weeks.

michdolan commented on vendor/openfx/OCIOColorSpace.cpp:
```diff
@@ -0,0 +1,224 @@
+    desc.addSupportedContext(OFX::eContextGeneral);
+
+    // Supported pixel depths
+    desc.addSupportedBitDepth(OFX::eBitDepthUByte);
```
OFX supports uint8 uint16, float16, and float32 via an enum. It does have a special value for a "custom" bit-depth, but I have not investigated how to handle that yet. It would appear however that Flame is converting the 10 and 12 bit inputs to float to comply with the plugin's declared support for these four bit-depths, which in reading the OFX docs, is correct host behavior. The plugin is also currently declaring that it doesn't support multiple bit-depths, and I'm creating an optimized CPU processor with matching input and output depths. Perhaps I should drop support of the integer bit depths in the plugin declaration, which would enforce float processing, and leave it up to the host to convert that back to an integer if needed. I know that would limit the integer input/output optimization in OCIO, but it would also guarantee color space precision from of the plugin.

doug-walker replied:
I would guess that most of the usage of these plug-ins will have float on one side or the other or both.  Also, most hosts probably either don't support kOfxImageEffectPropSupportsMultipleClipDepths or don't give users easy control for managing the depths.  So I would concur that we should probably just limit support to halfs and floats.  This will be a performance hit for apps that might like to pass in integers but we will typically not want to return an integer anyway.

michdolan replied:
Sounds good. Changed as discussed.

michdolan commented on vendor/openfx/OCIOProcessor.cpp:
```diff
@@ -0,0 +1,115 @@
+{
+    _srcImg = img;
+
+    // Make sure input and output bit-depth match
```
This check really isn't needed. By default plugins support only one bit-depth for input and output. In general the host and plugin declare what they support, and the host is responsible for providing a compatible input and output buffer. I'll remove this check and put some comments to clarify.

michdolan replied:
https://openfx.readthedocs.io/en/master/Reference/ofxClipPreferences.html

michdolan commented on vendor/openfx/OCIOColorSpace.cpp:
```diff
@@ -0,0 +1,224 @@
+    page->addChild(*swapSrcDstParam);
+
+    // Context overrides
+    defineContextParams(desc, page);
```
The reason the context values are blank by default is that they will only override the context if they are non-empty, otherwise the default OCIO-evaluated value is used. Nuke uses a checkbox to toggle overrides, but I was attempting to simplify it here. If I were to set the value with the current context variable value by default, it would bake the current env state into the plugin, which may not be desired, since it would be persistent.

As for the labels not appearing, I'm curious if Flame is failing to display the labels because they are within a (collapsible) group parameter. Perhaps an issue in the host implementation? The labels are set the same as for all other parameters and appear in Resolve and Natron. That's the only clear difference I can see in terms of the code.

michdolan commented:
> When I use a config that has a different set of views for each display, I'm noticing that when I change the display, the view menu doesn't update with the appropriate new values. It still has the old values. Does it update correctly in the other hosts you tested in?

I just tested this again in Resolve, and the views do change per display. The change is triggered by the OFX plugin `changedParam` callback. Perhaps Flame is not emitting that?

doug-walker commented:
Flame is definitely calling the changedParam callback because I'm getting the warning dialog box for the ColorSpace plug-in.  But I'll explore more why the menu is not being updated with the new values.  Thanks Michael!

hodoulp commented:
@michdolan & @doug-walker . Looking at the comments that's tricky to see if the pull request is ready or not. So, what is the status of the pull request?

doug-walker commented:
@hodoulp, I think Michael said he will be making another commit after his vacation.

michdolan commented on vendor/openfx/OCIOColorSpace.cpp:
```diff
@@ -0,0 +1,224 @@
+            configName = " '" + configName + "'";
+        }
+
+        std::string defaultViewTrName(config->getDefaultViewTransformName());
```
Thanks. Fixed.

michdolan commented on vendor/openfx/OCIOColorSpace.cpp:
```diff
@@ -0,0 +1,224 @@
+        {
+            std::ostringstream os;
+            os << "Color space '" << srcCsName << "' is ";
+            os << (srcRef == OCIO::REFERENCE_SPACE_SCENE ? "scene" : "display");
```
Good call. Changed.

michdolan commented on vendor/openfx/OCIOColorSpace.cpp:
```diff
@@ -0,0 +1,224 @@
+            os << numViewTransforms << " view transforms and the default '";
+            os << defaultViewTrName << "' will be used for this conversion, ";
+            os << "which may not be the intended behvior. Please use ";
+            os << "'OCIODisplay' to ensure a correct viewing pipeline result.";
```
Changed as suggested.

michdolan commented on vendor/openfx/OCIOColorSpace.cpp:
```diff
@@ -0,0 +1,224 @@
+    OFX::ChoiceParamDescriptor * srcCsNameParam = defineCsNameParam(
+        desc, 
+        "src_cs", 
+        "Src Color Space", 
```
Changed as suggested.

michdolan commented on vendor/openfx/OCIOColorSpace.cpp:
```diff
@@ -0,0 +1,224 @@
+        desc, 
+        "swap_src_dst", 
+        "Swap color spaces", 
+        "Sap src and dst color spaces",
```
Good catch. Thanks! Fixed.

michdolan commented on vendor/openfx/OCIODisplay.cpp:
```diff
@@ -0,0 +1,184 @@
+void OCIODisplayFactory::describe(OFX::ImageEffectDescriptor& desc)
+{
+    // Labels
+    desc.setLabels("OCIODisplay", "OCIODisplay", "OCIODisplay");
```
Good idea. Changed.

michdolan commented on vendor/openfx/OCIODisplay.cpp:
```diff
@@ -0,0 +1,184 @@
+    OFX::ChoiceParamDescriptor * srcCsNameParam = defineCsNameParam(
+        desc, 
+        "src_cs", 
+        "Src Color Space", 
```
Changed.

michdolan commented on vendor/openfx/OCIOProcessor.cpp:
```diff
@@ -0,0 +1,115 @@
+                                 OCIO::TransformDirection direction)
+{
+    OCIO::ConstConfigRcPtr config = getOCIOConfig();
+    OCIO::BitDepth bitDepth = getOCIOBitDepth(_dstImg->getPixelDepth());
```
Changed for clarity.

michdolan commented:
I just pushed commits which respond to @doug-walker 's feedback, and also adds support for preserving (and warning about) missing color spaces, displays, and views. The missing values are added to the menus, with a "(missing)" label for hosts that support it (Resolve doesn't seem to), and the preserved values are restored when the correct OCIO config is loaded in a future session.

michdolan commented:
Other than responding to additional review comments, or fixing CI issues that may occur, this PR is complete. I plan to follow up with additional plugins and improvements in later PRs.

michdolan commented:
@jmertic EasyCLA is remaining at `Expected - Waiting for status to be reported` for this PR. Any idea as to what the issue might be? This PR did exist prior to the CLA switch. Thanks!

doug-walker commented on vendor/openfx/OCIOUtils.cpp:
```diff
@@ -0,0 +1,651 @@
+
+    OCIO::ConstConfigRcPtr config = getOCIOConfig();
+
+    // Populate color space names
```
To the extent that these plug-ins will serve as examples to app developers, it would be nice to encourage people to use ColorSpaceMenuHelper.  Perhaps add something like: "TODO: Use ColorSpaceMenuHelper to generate the menus in order to leverage features such as categories."

michdolan replied:
Added.

doug-walker reviewed (approved):
Thanks for adding the mechanism to store a string for the color space names.  Apps that store only an integer are fragile since the config may change over the course of a project, so I like what you've done.  I have not had a chance to test your latest changes in a host app, but we could go ahead and merge this as soon as you're ready and do any follow-ups later, as needed.

michdolan commented:
I'm going to close and reopen this PR to try and trigger the EasyCLA check.

WillsonHG commented:
/easycla

WillsonHG commented:
/easycla

michdolan commented:
/easycla

michdolan commented:
Note that I added info on `OFX_PLUGIN_PATH` to the OCIO docs, which should point to the OFX bundle, along with the shared OCIO lib being on `PATH` to use these. That's how I test the plugins on Windows.

hodoulp commented on vendor/openfx/OCIOProcessor.h:
```diff
@@ -0,0 +1,41 @@
+    OCIOProcessor(OFX::ImageEffect & instance)
+        : OFX::ImageProcessor(instance)
+        , _srcImg(0)
+        , _cpuProc(nullptr)
```
[Minor] Smart pointers do not need to be initialized with nullptr. That's done.
In C++11 or higher, the naked pointer initialization should use `nullptr` instead of `0`.

hodoulp replied:
.. in fact, you could change line 17 to `OFX::Image * _srcImg = nullptr;`

michdolan replied:
Fixed. Thanks!

hodoulp reviewed (approved):
Only minor C++11 or higher code writing comments.
